examples/pytorch/schnet/layers.py

```diff
@@ -0,0 +1,302 @@
+        super().__init__(beta, threshold)
+        self.shift = shift
+        self.softplus = Softplus(beta, threshold)
+
+    def forward(self, input):
+        return self.softplus(input) - np.log(float(self.shift))
+
+
+class RBFLayer(nn.Module):
+    """
+    Radial basis functions Layer.
+    e(d) = exp(- gamma * ||d - mu_k||^2)
+    default settings:
+        gamma = 10
+        0 <= mu_k <= 30 for k=1~300
+    """
+
+    def __init__(self, low=0, high=30, gap=0.1, dim=1):
+        super().__init__()
+        self._low = low
+        self._high = high
+        self._gap = gap
+        self._dim = dim
+
+        self._n_centers = int(np.ceil((high - low) / gap))
+        centers = np.linspace(low, high, self._n_centers)
+        self.centers = th.tensor(centers, dtype=th.float, requires_grad=False)
+        self.centers = nn.Parameter(self.centers, requires_grad=False)
+        self._fan_out = self._dim * self._n_centers
+
+        self._gap = centers[1] - centers[0]
+
+    def dis2rbf(self, edges):
+        dist = edges.data["distance"]
+        radial = dist - self.centers
+        coef = -1 / self._gap
+        rbf = th.exp(coef * (radial**2))
+        return {"rbf": rbf}
+
+    def forward(self, g):
+        """Convert distance scalr to rbf vector"""
+        g.apply_edges(self.dis2rbf)
+        return g.edata["rbf"]
+
+
+class CFConv(nn.Module):
+    """
+    The continuous-filter convolution layer in SchNet.
+    One CFConv contains one rbf layer and three linear layer
+        (two of them have activation funct).
+    """
+
+    def __init__(self, rbf_dim, dim=64, act="sp"):
+        """
+        Args:
+            dim: the dimension of linear layers
+            act: activation function (default shifted softplus)
+        """
+        super().__init__()
+        self._rbf_dim = rbf_dim
+        self._dim = dim
+
+        self.linear_layer1 = nn.Linear(self._rbf_dim, self._dim)
+        self.linear_layer2 = nn.Linear(self._dim, self._dim)
+
+        if act == "sp":
+            self.activation = nn.Softplus(beta=0.5, threshold=14)
+        else:
+            self.activation = act
+
+    def update_edge(self, edges):
+        rbf = edges.data["rbf"]
+        h = self.linear_layer1(rbf)
+        h = self.activation(h)
+        h = self.linear_layer2(h)
+        # h = self.activation(h)
+        return {"h": h}
+
+    def send_edge_info(self, edges):
+        update_info = edges.src["new_node"] * edges.data["h"]
+        return {"neighbor_info": update_info}
+
+    def elemnet_wise_gather(self, nodes):
+        """
+        Implement element-wise multiplication (node x edge)
+        and reduce (sum) to nodes.
+        """
+        new_node = th.sum(nodes.mailbox["neighbor_info"], dim=1)
+        return {"new_node": new_node}
+
+    def forward(self, g):
+        g.apply_edges(self.update_edge)
+        g.send_and_recv(g.edges(),
+                        message_func=self.send_edge_info,
```

Comment:
Hi, I think the combination of your message function and reduce function could be optimized via our built-in function routine, which is much faster and memory efficient: https://docs.dgl.ai/features/builtin.html.
```suggestion
                        message_func=fn.u_mul_e('new_node', 'h', 'neighbor_info'),
                        reduce_func=fn.sum('neighbor_info', 'new_node'))
```

and don't forget to `import dgl.function as fn`.